src/extensions/id.rs

```diff
@@ -0,0 +1,43 @@
+//! IMAP ID extension specified in [RFC2971](https://datatracker.ietf.org/doc/html/rfc2971)
+
+use async_channel as channel;
+use futures::io;
+use futures::prelude::*;
+use imap_proto::{self, RequestId, Response};
+use std::collections::HashMap;
+
+use crate::types::ResponseData;
+use crate::types::*;
+use crate::{
+    error::Result,
+    parse::{filter, handle_unilateral},
+};
+
+pub(crate) async fn parse_id<T: Stream<Item = io::Result<ResponseData>> + Unpin>(
+    stream: &mut T,
+    unsolicited: channel::Sender<UnsolicitedResponse>,
+    command_tag: RequestId,
+) -> Result<Option<HashMap<String, String>>> {
+    let mut id = None;
+    while let Some(resp) = stream
+        .take_while(|res| filter(res, &command_tag))
+        .next()
+        .await
+    {
+        let resp = resp?;
+        match resp.parsed() {
+            Response::Id(res) => {
+                id = res.as_ref().map(|m| {
+                    m.iter()
+                        .map(|(k, v)| (k.to_string(), v.to_string()))
```

Comment:
I would suggest adding some validation here

>   Field strings MUST NOT be longer than 30 octets.  Value strings MUST
   NOT be longer than 1024 octets.  Implementations MUST NOT send more
   than 30 field-value pairs.

Reply:
👍